internal/crypto/cert.go

```diff
@@ -194,65 +97,46 @@ func (ca *CA) MakeAndWriteServerCert(certFile, keyFile string, hostnames []strin
 	if len(hostnames) < 1 {
 		return nil, fmt.Errorf("at least one hostname must be provided")
 	}
 
-	serverPublicKey, serverPrivateKey, publicKeyHash, _ := NewKeyPairWithHash()
+	_, serverPrivateKey, _ := NewKeyPair()
 
-	now := time.Now()
-	serverTemplate := &x509.Certificate{
+	serverTemplate := &x509.CertificateRequest{
 		Subject: pkix.Name{CommonName: hostnames[0]},
-
-		SignatureAlgorithm: x509.ECDSAWithSHA256,
-
-		NotBefore:    now.Add(-1 * time.Second),
-		NotAfter:     now.Add(time.Duration(expiryDays) * 24 * time.Hour),
-		SerialNumber: big.NewInt(1),
-
-		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature,
-		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
-		BasicConstraintsValid: true,
-
-		AuthorityKeyId: ca.Config.Certs[0].SubjectKeyId,
-		SubjectKeyId:   publicKeyHash,
 	}
 	serverTemplate.IPAddresses, serverTemplate.DNSNames = oscrypto.IPAddressesDNSNames(hostnames)
-	for _, fn := range fns {
-		if err := fn(serverTemplate); err != nil {
-			return nil, err
-		}
-	}
 
-	serverCrt, err := ca.signCertificate(serverTemplate, serverPublicKey)
+	raw, err := x509.CreateCertificateRequest(rand.Reader, serverTemplate, serverPrivateKey)
 	if err != nil {
 		return nil, err
 	}
-	server := &TLSCertificateConfig{
-		Certs: append([]*x509.Certificate{serverCrt}, ca.Config.Certs...),
-		Key:   serverPrivateKey,
+	csr, err := x509.ParseCertificateRequest(raw)
+	if err != nil {
+		return nil, err
 	}
-	return server, nil
-}
-
-func (ca *CA) signCertificate(template *x509.Certificate, requestKey crypto.PublicKey) (*x509.Certificate, error) {
-	// Increment and persist serial
-	serial, err := ca.SerialGenerator.Next(template)
+	serverCrt, err := caClient.IssueRequestedServerCertificateAsX509(csr, expiryDays*86400)
 	if err != nil {
 		return nil, err
 	}
-	template.SerialNumber = big.NewInt(serial)
-	return signCertificate(template, requestKey, ca.Config.Certs[0], ca.Config.Key)
+	server := &TLSCertificateConfig{
+		Certs: append([]*x509.Certificate{serverCrt}, caClient.GetCABundleX509()...),
+		Key:   serverPrivateKey,
+	}
+	return server, nil
 }
 
-func (ca *CA) EnsureClientCertificate(certFile, keyFile string, subjectName string, expireDays int) (*TLSCertificateConfig, bool, error) {
+func (caClient *CA) EnsureClientCertificate(certFile, keyFile string, subjectName string, expireDays int) (*TLSCertificateConfig, bool, error) {
 	certConfig, err := GetClientCertificate(certFile, keyFile, subjectName)
 	if err != nil {
-		certConfig, err = ca.MakeClientCertificate(certFile, keyFile, subjectName, expireDays)
-		return certConfig, true, err // true indicates we wrote the files.
+		certConfig, err := caClient.MakeClientCert(certFile, keyFile, subjectName, expireDays)
+		if err != nil {
+			return nil, false, err
+		}
+		err = certConfig.WriteCertConfigFile(certFile, keyFile)
+		return certConfig, err == nil, err // true indicates we wrote the files.
```

Comment:
isnt the presence of the error itself indicative of writing files why do we need also a bool

Reply:
See elsewhere - generated versus loaded.